flink/src/main/java/com/zhihu/flink/source/tidb/TiDBInputFormat.java

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.zhihu.flink.source.tidb;
+
+
+import com.zhihu.presto.tidb.*;
+import java.io.IOException;
+import java.time.LocalTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import org.apache.flink.api.common.io.DefaultInputSplitAssigner;
+import org.apache.flink.api.common.io.RichInputFormat;
+import org.apache.flink.api.common.io.statistics.BaseStatistics;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.io.GenericInputSplit;
+import org.apache.flink.core.io.InputSplit;
+import org.apache.flink.core.io.InputSplitAssigner;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.utils.TypeConversions;
+import org.apache.flink.types.Row;
+
+public class TiDBInputFormat extends RichInputFormat<Row, InputSplit> implements ResultTypeQueryable<Row> {
+
+  private String pdAddresses;
+
+  private String databaseName;
+
+  private String tableName;
+
+  private String[] fieldNames;
+
+  private DataType[] fieldTypes;
+
+  private List<SplitInternal> splits;
+
+  private List<ColumnHandleInternal> columnHandleInternals;
+
+  private transient RecordCursorInternal cursor;
+
+  private transient ClientSession clientSession;
+
+
+  /**
+   * see {@link TiDBInputFormat#builder()}
+   */
+  private TiDBInputFormat() {
+  }
+
+  @Override
+  public void configure(Configuration parameters) {
+    // do nothing here
+  }
+
+  @Override
+  public BaseStatistics getStatistics(BaseStatistics cachedStatistics) throws IOException {
+    return cachedStatistics;
+  }
+
+  @Override
+  public InputSplit[] createInputSplits(int minNumSplits) throws IOException {
+    GenericInputSplit[] inputSplits = new GenericInputSplit[splits.size()];
+    for (int i = 0; i < inputSplits.length; i++) {
+      inputSplits[i] = new GenericInputSplit(i, inputSplits.length);
+    }
+    return inputSplits;
+  }
+
+  @Override
+  public InputSplitAssigner getInputSplitAssigner(InputSplit[] inputSplits) {
+    return new DefaultInputSplitAssigner(inputSplits);
+  }
+
+  @Override
+  public void openInputFormat() throws IOException {
+    clientSession = new ClientSession(new ClientConfig(pdAddresses));
+  }
+
+  @Override
+  public void closeInputFormat() throws IOException {
+    // TODO clientSession will auto close?
```

Comment:
Presto connector use singleton ClientSession therefore doesn't require close. However it is not the case here. You can make ClientSession implement AutoCloseable and call close here.